cmd/xdpcap/filter.go

```diff
@@ -85,13 +85,26 @@ func newFilterWithMap(hookMap *ebpf.Map, opts filterOpts) (*filter, error) {
 		actions:  opts.actions,
 	}
 
-	for _, action := range opts.actions {
-		program, err := newProgram(opts.filter, action, perfMap)
+	multiBufferMode := false
+	for i, action := range opts.actions {
+		program, err := newProgram(opts.filter, action, perfMap, multiBufferMode)
 		if err != nil {
 			return nil, errors.Wrapf(err, "loading filter program for %v", action)
 		}
 
 		err = attachProg(hookMap, program.program.FD(), action)
+		if err != nil && i == 0 {
+			// attempt to load first action in multi-buffer mode and retry attaching
+			// if this doesn't work then there is some other error and the multi-buffer
+			// mode is not the culprit
+			multiBufferMode = true
+			program, err = newProgram(opts.filter, action, perfMap, multiBufferMode)
```

Comment:
fixed

Reply:
By "report the original error", I meant store the new error in a different variable, and return the original one if using `xdp.frags` didn't fix anything:

```
		err = attachProg(hookMap, program.program.FD(), action)
		if errors.Is(err, unix.EINVAL) && i == 0 {
			multiBufferMode = true
                        var fragsErr error
			program, fragsErr = newProgram(opts.filter, action, perfMap, multiBufferMode)
                        // If using XDP frags didn't help, the original error was legitimate.
                        // Return it so we don't confuse the user with a potentially unrelated frags error.
                        if fragsErr != nil {
                            return err
                        }
```

None of the rest of the code prints anything or logs errors here (we only print errors at the top level: https://github.com/cloudflare/xdpcap/blob/master/cmd/xdpcap/main.go#L51). That makes it easier to test, and avoids us ending up with lots of verbose output. Can we drop the prints?

Reply:
I opted to use the new `errors.Join` function over here that allows me to join both errors into one if the attach operation doesn't succeed at all. I think this would be more useful to the user for debugging purposes.
I get why you don't want to print the first attempt error, but when it comes to loading in multi buffer mode I think it would be useful information for the user to be aware that the tool attached in multi-buffer mode

Reply:
I think printing "setting up multi buffer mode for filter actions" and printing both errors is more confusing - to understand what's really happening you have to look at the code anyways. But we've gone back and forth on this enough, and it's just a detail.

Reply:
I modified this to keep only the first error around and return it if the second attempt at attaching also fails. I also changed the log message to only be printed if the second attempt using `xdp.frags` mode works.